napari/_vispy/vispy_image_layer.py

```diff
@@ -74,6 +76,24 @@ def _on_data_change(self):
         if self.layer.dims.ndisplay == 3 and self.layer.dims.ndim == 2:
             data = np.expand_dims(data, axis=0)
 
+        if self.MAX_TEXTURE_SIZE is not None:
+            if np.any(np.greater(data.shape, self.MAX_TEXTURE_SIZE)):
+                warnings.warn(
+                    f"data shape {data.shape} exceeds GL_MAX_TEXTURE_SIZE"
+                    f"{self.MAX_TEXTURE_SIZE} in at least one axis and will be "
+                    f"downsampled."
+                )
+                downsample = np.ceil(
+                    np.divide(data.shape, self.MAX_TEXTURE_SIZE)
+                ).astype(int)
+                scale = np.ones(self.layer.ndim)
+                for i, d in enumerate(self.layer.dims.displayed):
+                    scale[d] = downsample[i]
```

Comment:
This seems like it would break the existing scale that a user might have set? Perhaps multiply by downsample[i]?

Reply:
(And this is the sort of thing that I meant when I said that doing this correctly is difficult.)

Reply:
Totally agree on the complexity here, this will be ok as long as the data is not also a pyrmaid!!! `_scale_view` is a separate scale factor that we use the pyramid stuff that I'm leveraging here too. I could add an explicit step which throws an error if you are a pyramid and your tile size is too big!!!

Reply:
I've added a catch in case a pyramid ever fails this test - but that is something entirely within our hands - as we control `layer._max_tile_shape` currently `1600`. To see why we don't need the multiplication here, it happens bellow:
https://github.com/napari/napari/blob/67789fba557ae58436437f4c8e3c123c5fbab22c/napari/_vispy/vispy_image_layer.py#L117
where the user provided `scale` interacts with our `_scale_view` that can handle additional offsets. It should work fine, but agree this could have hidden bugs in edge cases